Shell/Shell.cpp

```diff
@@ -571,10 +566,27 @@ RefPtr<Job> Shell::run_command(const AST::Command& command)
         ASSERT_NOT_REACHED();
     }
 
+    bool is_first = !command.pipeline || (command.pipeline && command.pipeline->pgid == -1);
+
+    if (command.pipeline) {
+        if (is_first) {
+            command.pipeline->pgid = child;
+        }
+    }
+
+    pid_t pgid = is_first ? child : (command.pipeline ? command.pipeline->pgid : child);
+    if (setpgid(child, pgid) < 0)
+        perror("setpgid");
+
+    if (command.should_wait) {
+        tcsetpgrp(STDOUT_FILENO, pgid);
+        tcsetpgrp(STDIN_FILENO, pgid);
+    }
+
     StringBuilder cmd;
     cmd.join(" ", command.argv);
 
-    auto job = Job::create(child, (unsigned)child, cmd.build(), find_last_job_id() + 1);
+    auto job = Job::create(child, pgid, cmd.build(), find_last_job_id() + 1, command.pipeline);
```

Comment:
_Something_ isn't quite right here, as I see some stack corruptions here and there, but other than that, I think this is a pretty ok way of dealing with pipelines

Reply:
Now I'm truly perplexed, building with `-fno-omit-frame-pointer` makes the stack mishaps vanish into thin air
:whoosh:

I guess I'll leave it be for the time being until it resurfaces.